diff --git a/Makefile b/Makefile
index 4138da9..9f46b94 100644
--- a/Makefile
+++ b/Makefile
@@ -67,6 +67,7 @@ endif
 ifeq ($(ARCH), x86_64)
   # Don't enable kvm for WSL/WSL2.
   ACCEL ?= $(if $(findstring -microsoft, $(shell uname -r | tr '[:upper:]' '[:lower:]')),n,y)
+  ACCEL ?= n
   PLATFORM ?= x86_64-qemu-q35
   TARGET := x86_64-unknown-none
   BUS := pci
@@ -109,7 +110,11 @@ CROSS_COMPILE ?= $(ARCH)-linux-musl-
 CC := $(CROSS_COMPILE)gcc
 AR := $(CROSS_COMPILE)ar
 RANLIB := $(CROSS_COMPILE)ranlib
-LD := rust-lld -flavor gnu
+ifeq ($(ARCH), x86_64)
+	LD := rust-lld -flavor gnu
+else
+    LD := $(CROSS_COMPILE)ld
+endif
 
 OBJDUMP ?= rust-objdump -d --print-imm-hex --x86-asm-syntax=intel
 OBJCOPY ?= rust-objcopy --binary-architecture=$(ARCH)
diff --git a/apps/c/redis/redis b/apps/c/redis/redis
--- a/apps/c/redis/redis
+++ b/apps/c/redis/redis
@@ -1 +1 @@
-Subproject commit 8e73f9d34821a937165884f13a2981883f44a074
+Subproject commit 8e73f9d34821a937165884f13a2981883f44a074-dirty
diff --git a/modules/axhal/src/platform/x86_pc/misc.rs b/modules/axhal/src/platform/x86_pc/misc.rs
index 7af6231..6d88863 100644
--- a/modules/axhal/src/platform/x86_pc/misc.rs
+++ b/modules/axhal/src/platform/x86_pc/misc.rs
@@ -5,7 +5,17 @@ use x86_64::instructions::port::PortWriteOnly;
 /// See <https://wiki.osdev.org/Shutdown> for more information.
 pub fn terminate() -> ! {
     info!("Shutting down...");
-    unsafe { PortWriteOnly::new(0x604).write(0x2000u16) };
+
+    #[cfg(platform = "x86_64-pc-oslab")]
+    unsafe {
+        PortWriteOnly::new(0x64).write(0xfeu8)
+    };
+
+    #[cfg(platform = "x86_64-qemu-q35")]
+    unsafe {
+        PortWriteOnly::new(0x604).write(0x2000u16)
+    };
+
     crate::arch::halt();
     warn!("It should shutdown!");
     loop {
diff --git a/ulib/axlibc/c/env.c b/ulib/axlibc/c/env.c
index 131b2e1..a0505a9 100644
--- a/ulib/axlibc/c/env.c
+++ b/ulib/axlibc/c/env.c
@@ -3,7 +3,8 @@
 #include <string.h>
 #include <unistd.h>
 
-char **environ = 0;
+char *environ_[2] = {"dummy", NULL};
+char **environ = (char **)environ_;
 
 char *getenv(const char *name)
 {
diff --git a/ulib/axlibc/c/stdlib.c b/ulib/axlibc/c/stdlib.c
index 91c742d..8239087 100644
--- a/ulib/axlibc/c/stdlib.c
+++ b/ulib/axlibc/c/stdlib.c
@@ -9,8 +9,8 @@
 
 #include <axlibc.h>
 
-char *program_invocation_short_name = NULL;
-char *program_invocation_name = NULL;
+char *program_invocation_short_name = "dummy";
+char *program_invocation_name = "dummy";
 
 #define __DECONST(type, var) ((type)(uintptr_t)(const void *)(var))
 
diff --git a/ulib/axlibc/include/errno.h b/ulib/axlibc/include/errno.h
index f99d2c5..16ed78e 100644
--- a/ulib/axlibc/include/errno.h
+++ b/ulib/axlibc/include/errno.h
@@ -143,4 +143,8 @@
 int *__errno_location(void);
 #define errno (*__errno_location())
 
+#ifdef _GNU_SOURCE
+extern char *program_invocation_short_name, *program_invocation_name;
+#endif
+
 #endif // __ERRNO_H__
diff --git a/ulib/axlibc/include/math.h b/ulib/axlibc/include/math.h
index 8b6105b..649b851 100644
--- a/ulib/axlibc/include/math.h
+++ b/ulib/axlibc/include/math.h
@@ -13,6 +13,12 @@ typedef double double_t;
 #define INFINITY 1e5000f
 #endif
 
+#define M_PI 3.14159265358979323846 /* pi */
+
+#define LOG_TABLE_BITS  7
+#define LOG_POLY_ORDER  6
+#define LOG_POLY1_ORDER 12
+
 #define HUGE_VALF INFINITY
 #define HUGE_VAL  ((double)INFINITY)
 #define HUGE_VALL ((long double)INFINITY)
